FM-72 & FM-117: Add in module for assets and asset types

leeallen337 opened the pull request:
## Why?

https://ndustrialio.atlassian.net/browse/FM-72
https://ndustrialio.atlassian.net/browse/FM-117

Need modules for `Assets` and `Asset Types`

## What changed?
- Added `Assets` module
  - Added methods `create`, `delete`, `get`, `getAll`, `getAllByOrganizationId`, `update`
- Added `AssetTypes` module
  - Added methods `create`, `delete`, `get`, `getAll`, `getAllByOrganizationId`, `update`

leeallen337 commented:
I haven't bumped the version yet using `npm version`. Probably best to wait until it's been approved and ready to merge

greg5green commented on src/utils/assets/formatAssetFromServer.js:
```diff
@@ -0,0 +1,29 @@
+ * @param {string} input.id UUID
+ * @param {string} input.label
+ * @param {string} input.organization_id UUID
+ * @param {string} input.updated_at ISO 8601 Extended Format date/time string
```
Looking at this, I realize I need to tighten up the stuff I did last week in #41 in this regard. Which of these is optional?

leeallen337 replied:
`input.description` is optional. Updating 👍

greg5green commented on src/utils/assets/formatAssetOptionsToServer.js:
```diff
@@ -0,0 +1,25 @@
+ * Normalizes the options provided when retrieving assets
+ *
+ * @param {Object} options
+ * @param {string} options.assetTypeId ID of the asset type to use for filtering
```
This looks optional

greg5green commented on src/utils/assets/formatAssetToServer.js:
```diff
@@ -0,0 +1,23 @@
+ *
+ * @returns {Object} output
+ * @returns {string} output.asset_type_id UUID
+ * @returns {string} output.description
```
Optional?

leeallen337 replied:
Yep

greg5green commented on src/utils/assets/formatAssetTypeFromServer.js:
```diff
@@ -0,0 +1,27 @@
+ *
+ * @param {Object} input
+ * @param {string} input.created_at ISO 8601 Extended Format date/time string
+ * @param {string} input.description
```
Optional?

greg5green commented on src/utils/assets/formatAssetTypeToServer.js:
```diff
@@ -0,0 +1,21 @@
+ * @param {AssetType} input
+ *
+ * @returns {Object} output
+ * @returns {string} output.description
```
Optional?

leeallen337 replied:
Not for `AssetType`

greg5green commented on src/assets/assetTypes.spec.js:
```diff
@@ -0,0 +1,484 @@
+      let request;
+
+      beforeEach(function() {
+        assetTypeFromServerAfterFormat = fixture.build('assetType');
```
I like this naming scheme more than what I had been doing 👍

greg5green commented on src/assets/assetTypes.spec.js:
```diff
@@ -0,0 +1,484 @@
+        });
+        assetTypeFromServerBeforeFormat = fixture.build(
+          'assetType',
+          { id: expectedAssetTypeId },
```
You can make this a little simpler actually
```
assetTypeFromServerAfterFormat = fixture.build('assetType');
assetTypeFromServerBeforeFormat = fixture.build('assetType', assetTypeFromServerAfterFormat, { fromServer: true });
```

greg5green commented on src/assets/assetTypes.spec.js:
```diff
@@ -0,0 +1,484 @@
+    let assetTypesFromServerAfterFormat;
+    let assetTypesFromServerBeforeFormat;
+    let formatAssetTypeFromServer;
+    let numberofAssetTypes;
```
`numberOfAssetTypes`?

greg5green commented on src/assets/assetTypes.spec.js:
```diff
@@ -0,0 +1,484 @@
+      assetTypesFromServerBeforeFormat = fixture.buildList(
+        'assetType',
+        numberofAssetTypes,
+        { fromServer: true }
```
I think there needs to be a `null` here?

leeallen337 commented on CHANGELOG.md:
```diff
@@ -1,3 +1,22 @@
+  * AssetTypes#getAllByOrganizationId to get a list of all asset types for a specific organization
+  * AssetTypes#update to update an asset type
+
 ## [v0.0.16](http://github.com/ndustrialio/contxt-sdk-js/tree/v0.0.15)
```
Updated 👍

leeallen337 commented on src/assets/index.js:
```diff
@@ -0,0 +1,257 @@
+/**
+ * @typedef {Object} Asset
+ * @property {string} assetTypeId UUID corresponding with the asset type
+ * @property {string} description
```
`description` is optional. Updating 👍

leeallen337 commented on src/assets/index.js:
```diff
@@ -0,0 +1,257 @@
+   *
+   * @param {string} organizationId UUID corresponding with an organization
+   * @param {Object} options Object containing parameters to be called with the request
+   * @param {string} options.assetTypeId ID of the asset type to use for filtering
```
They are both optional. Updating 👍

leeallen337 commented on src/assets/index.js:
```diff
@@ -0,0 +1,257 @@
+   *
+   * @param {string} assetId The ID of the asset to update (formatted as a UUID)
+   * @param {Object} update An object containing the updated data for the asset
+   * @param {string} update.description
```
Yep

leeallen337 commented on src/assets/assetTypes.js:
```diff
@@ -0,0 +1,251 @@
+   *
+   * @param {string} assetTypeId The ID of the asset type to update (formatted as a UUID)
+   * @param {Object} update An object containing the updated data for the asset type
+   * @param {string} update.description
```
I don't think we're going to allow them to update `label` for now. See [here](https://github.com/sustainableis/facilities-service/pull/36#discussion_r196870704)

leeallen337 commented on src/assets/assetTypes.js:
```diff
@@ -0,0 +1,251 @@
+
+/**
+ * @typedef {Object} AssetType
+ * @property {string} description
```
Currently. `description` for an `Asset` is optional but it is required for an `AssetType`. See the [required fields](https://github.com/sustainableis/facilities-service/blob/master/controllers/asset_types.js#L12). I think all fields are required for `AssetType`

leeallen337 commented on src/assets/assetTypes.js:
```diff
@@ -0,0 +1,251 @@
+      .get(`${this._baseUrl}/assets/types`)
+      .then((response) => {
+        return {
+          ...response,
```
Updated with a formatter

leeallen337 commented on src/assets/assetTypes.js:
```diff
@@ -0,0 +1,251 @@
+   * Method: GET
+   *
+   * @returns {Promise}
+   * @fulfill {AssetType[]} Information about all asset types
```
Updated with a type

leeallen337 commented on src/assets/assetTypes.js:
```diff
@@ -0,0 +1,251 @@
+      .get(`${this._baseUrl}/organizations/${organizationId}/assets/types`)
+      .then((response) => {
+        return {
+          ...response,
```
Updated with a formatter

leeallen337 commented on src/assets/assetTypes.js:
```diff
@@ -0,0 +1,251 @@
+   * @param {string} organizationId UUID corresponding with an organization
+   *
+   * @returns {Promise}
+   * @fulfill {AssetType[]} Information about all asset types
```
Updated with a type

leeallen337 commented on src/assets/index.js:
```diff
@@ -0,0 +1,257 @@
+   * Method: GET
+   *
+   * @returns {Promise}
+   * @fulfill {Asset[]} Information about all assets
```
Updated with a type

leeallen337 commented on src/assets/index.js:
```diff
@@ -0,0 +1,257 @@
+  getAll() {
+    return this._request.get(`${this._baseUrl}/assets`).then((response) => {
+      return {
+        ...response,
```
Updated with a formatter

leeallen337 commented on src/assets/index.js:
```diff
@@ -0,0 +1,257 @@
+      })
+      .then((response) => {
+        return {
+          ...response,
```
Updated with a formatter

leeallen337 commented:
@greg5green I believe I've addressed all your comments. I added in a formatter for the pagination data and some new specs and updates the way the `getAll` and `getAllByOrganizationId` methods were being handled. Whenever you have time it could use another review
